File: android/expoview/src/main/java/abi26_0_0/host/exp/exponent/modules/api/MailComposerModule.java
package abi26_0_0.host.exp.exponent.modules.api;

import android.app.Activity;
import android.content.ComponentName;
import android.content.Intent;
import android.content.pm.LabeledIntent;
import android.content.pm.ResolveInfo;
import android.net.Uri;
import android.text.Html;
import android.text.Spanned;

import abi26_0_0.com.facebook.react.bridge.Arguments;
import abi26_0_0.com.facebook.react.bridge.LifecycleEventListener;
import abi26_0_0.com.facebook.react.bridge.Promise;
import abi26_0_0.com.facebook.react.bridge.ReactApplicationContext;
import abi26_0_0.com.facebook.react.bridge.ReactContextBaseJavaModule;
import abi26_0_0.com.facebook.react.bridge.ReactMethod;
import abi26_0_0.com.facebook.react.bridge.ReadableArray;
import abi26_0_0.com.facebook.react.bridge.ReadableMap;
import abi26_0_0.com.facebook.react.bridge.WritableMap;

import java.io.File;
import java.util.ArrayList;
import java.util.List;

import host.exp.exponent.utils.ExpFileUtils;

public class MailComposerModule extends ReactContextBaseJavaModule implements LifecycleEventListener {

  private boolean mComposerOpened = false;
  private Promise mPromise;

  public MailComposerModule(ReactApplicationContext reactContext) {
    super(reactContext);
    getReactApplicationContext().addLifecycleEventListener(this);
  }

  @Override
  public String getName() {
    return "ExponentMailComposer";
  }

  @ReactMethod
  public void composeAsync(ReadableMap options, Promise promise) {

    Intent intent = new Intent(Intent.ACTION_SENDTO);
    intent.setData(Uri.parse("mailto:"));

    String[] recipients = {}, ccRecipients = {}, bccRecipients = {};
    String subject = "";
    String body = "";
    Spanned htmlBody = null;
    ArrayList<Uri> attachments = new ArrayList<>();

    if (options.hasKey("recipients")) {
      recipients =  options.getArray("recipients").toArrayList().toArray(new String[0]);
    }

    if (options.hasKey("ccRecipients")) {
      ccRecipients = options.getArray("ccRecipients").toArrayList().toArray(new String[0]);
    }

    if (options.hasKey("bccRecipients")) {
      bccRecipients =  options.getArray("bccRecipients").toArrayList().toArray(new String[0]);
    }

    if (options.hasKey("subject")) {
      subject = options.getString("subject");
    }

    boolean isHtml = options.hasKey("isHtml") && options.getBoolean("isHtml");
    if (options.hasKey("body")) {
      if (isHtml) {
        htmlBody = Html.fromHtml(options.getString("body"));
      } else {
        body = options.getString("body");
      }
    }

    if (options.hasKey("attachments")) {
      ReadableArray requestedAttachments = options.getArray("attachments");
      for (int i = 0; i < requestedAttachments.size(); i++) {
        File attachmentFile = new File(Uri.parse(options.getArray("attachments").getString(i)).getPath());
        attachments.add(ExpFileUtils.contentUriFromFile(attachmentFile));
      }
    }

    List<ResolveInfo> resolveInfos = getReactApplicationContext().getPackageManager().queryIntentActivities(intent, 0);
    List<LabeledIntent> mailIntents = new ArrayList<>();
    for (ResolveInfo info : resolveInfos) {
      Intent mailIntent = new Intent(Intent.ACTION_SEND_MULTIPLE);
      mailIntent.setComponent(new ComponentName(info.activityInfo.packageName, info.activityInfo.name));
      mailIntent.putExtra(Intent.EXTRA_EMAIL, recipients);
      mailIntent.putExtra(Intent.EXTRA_CC, ccRecipients);
      mailIntent.putExtra(Intent.EXTRA_BCC, bccRecipients);
      mailIntent.putExtra(Intent.EXTRA_SUBJECT, subject);
      mailIntent.putExtra(Intent.EXTRA_TEXT, isHtml ? htmlBody : body);
      mailIntent.putParcelableArrayListExtra(Intent.EXTRA_STREAM, attachments);
      mailIntents.add(new LabeledIntent(mailIntent, info.activityInfo.packageName, info.loadLabel(getReactApplicationContext().getPackageManager()), info.icon));
    }
    Intent chooser = Intent.createChooser(mailIntents.remove(mailIntents.size() - 1), null);
    chooser.putExtra(Intent.EXTRA_INITIAL_INTENTS, mailIntents.toArray(new LabeledIntent[mailIntents.size()]));

    mPromise = promise;
    chooser.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
    chooser.addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
    getReactApplicationContext().startActivity(chooser);
    mComposerOpened = true;
  }

  @Override
  public void onHostResume() {
    if (mComposerOpened) {
      mComposerOpened = false;
      WritableMap response = Arguments.createMap();
      response.putString("status", "sent");
      mPromise.resolve(response);
    }
  }

  @Override
  public void onHostPause() {

  }

  @Override
  public void onHostDestroy() {

  }
}

File: android/versioned-abis/expoview-abi48_0_0/src/main/java/abi48_0_0/expo/modules/mailcomposer/MailComposerModule.kt
package abi48_0_0.expo.modules.mailcomposer

import android.app.Activity
import android.content.Context
import android.content.Intent
import android.content.pm.LabeledIntent
import android.net.Uri
import android.os.Bundle
import abi48_0_0.expo.modules.core.ExportedModule
import abi48_0_0.expo.modules.core.ModuleRegistry
import abi48_0_0.expo.modules.core.ModuleRegistryDelegate
import abi48_0_0.expo.modules.core.Promise
import abi48_0_0.expo.modules.core.arguments.ReadableArguments
import abi48_0_0.expo.modules.core.interfaces.ActivityProvider
import abi48_0_0.expo.modules.core.interfaces.ExpoMethod
import abi48_0_0.expo.modules.core.interfaces.ActivityEventListener
import abi48_0_0.expo.modules.core.interfaces.services.UIManager

class MailComposerModule(
  context: Context,
  private val moduleRegistryDelegate: ModuleRegistryDelegate = ModuleRegistryDelegate()
) : ExportedModule(context), ActivityEventListener {
  private var composerOpened = false
  private val uiManager: UIManager by moduleRegistry()
  private var pendingPromise: Promise? = null
  override fun getName() = "ExpoMailComposer"
  private val activityProvider: ActivityProvider by moduleRegistry()

  private inline fun <reified T> moduleRegistry() = moduleRegistryDelegate.getFromModuleRegistry<T>()

  override fun onCreate(moduleRegistry: ModuleRegistry) {
    moduleRegistryDelegate.onCreate(moduleRegistry)
    uiManager.registerActivityEventListener(this)
  }

  override fun onDestroy() {
    uiManager.unregisterActivityEventListener(this)
  }

  @ExpoMethod
  fun isAvailableAsync(promise: Promise) {
    promise.resolve(true)
  }

  @ExpoMethod
  fun composeAsync(options: ReadableArguments, promise: Promise) {
    val intent = Intent(Intent.ACTION_SENDTO).apply { data = Uri.parse("mailto:") }
    val application = activityProvider.currentActivity.application
    val resolveInfo = context.packageManager.queryIntentActivities(intent, 0)
    val mailIntents = resolveInfo.map { info ->
      val isHtml = options.containsKey("isHtml") && options.getBoolean("isHtml")
      val mailIntentBuilder = MailIntentBuilder(options)
        .setComponentName(info.activityInfo.packageName, info.activityInfo.name)
        .putExtraIfKeyExists("recipients", Intent.EXTRA_EMAIL)
        .putExtraIfKeyExists("ccRecipients", Intent.EXTRA_CC)
        .putExtraIfKeyExists("bccRecipients", Intent.EXTRA_BCC)
        .putExtraIfKeyExists("subject", Intent.EXTRA_SUBJECT)
        .putExtraIfKeyExists("body", Intent.EXTRA_TEXT, isHtml)
        .putParcelableArrayListExtraIfKeyExists(
          "attachments",
          Intent.EXTRA_STREAM,
          application
        )
      LabeledIntent(
        mailIntentBuilder.build(),
        info.activityInfo.packageName,
        info.loadLabel(context.packageManager),
        info.icon
      )
    }.toMutableList()
    val chooser = Intent.createChooser(
      mailIntents.removeAt(mailIntents.size - 1),
      null
    ).apply {
      putExtra(Intent.EXTRA_INITIAL_INTENTS, mailIntents.toTypedArray())
      addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
    }
    pendingPromise = promise
    activityProvider.currentActivity.startActivityForResult(chooser, REQUEST_CODE)
    composerOpened = true
  }

  override fun onNewIntent(intent: Intent) = Unit

  override fun onActivityResult(activity: Activity, requestCode: Int, resultCode: Int, data: Intent?) {
    if (requestCode == REQUEST_CODE && pendingPromise != null) {
      val promise = pendingPromise ?: return
      if (composerOpened) {
        composerOpened = false
        promise.resolve(Bundle().apply { putString("status", "sent") })
      }
    }
  }

  companion object {
    private const val REQUEST_CODE = 8675
  }
}
